src/async-components/views/dialogs/security/CreateSecretStorageDialog.tsx

```diff
@@ -382,6 +382,7 @@ export default class CreateSecretStorageDialog extends React.PureComponent<IProp
                 //     moment via token login)
                 await crypto.bootstrapCrossSigning({
                     authUploadDeviceSigningKeys: this.doBootstrapUIAuth,
+                    setupNewCrossSigning: true,
```

Comment:
Isn't it worth it to introduce extra checks to ensure that the value is just set in the described setting?
Like:
```typescript
await crypto.bootstrapCrossSigning({
    authUploadDeviceSigningKeys: this.doBootstrapUIAuth,
    setupNewCrossSigning: setupNewCrossSigning:  !(await cli.secretStorage.hasKey()) ? true : undefined
});
```
The above code would maybe still cause problems for users who were deactivated (with purge) and reactivated.

This bug prevents me to update my system because the messages of new users would be lost / not decryptable after signing out and in again.

Reply:
> Isn't it worth it to introduce extra checks to ensure that the value is just set in the described setting? Like:
> 
> ```ts
>     setupNewCrossSigning: setupNewCrossSigning:  !(await cli.secretStorage.hasKey()) ? true : undefined
> ```

The logic here isn't very clear, but I think this means: "create new cross-signing keys if there is no master secret-storage key". But according to your [bug report](https://github.com/element-hq/element-web/issues/27382#issuecomment-2071623521), you *do* have a master secret-storage key (`account_data` contains `m.secret_storage.default_key` and a `m.secret_storage.key.*`). So this would do nothing to help in your situation?

I still don't understand why we need to replace existing cross-signing keys here.

In any case, if this logic is needed, it would be better placed the `bootstrapCrossSigning` implementation than here.

Reply:
I agree, that ```bootstrapCrossSigning``` seems to be the correct place. 

I don't know the details about the secret storage and the cross signing keys, but from the code, debugging and the database entries it looks like the ```bootstrapCrossSigning``` process does not require the secret storage to be setup (no specific account_data entries) before taking actions. In the issue is described that no master secret-storage key exists when the process starts (very first interaction with the chat for this user). Also ```CreateSecretStorage.bootstrapSecretStorage``` first invokes
```crypto.bootstrapCrossSigning(....``` before ```crypto.bootstrapSecretStorage((...```
I know that know the implementation waayy better than me, so I wrote this just to make clear what my thoughts are.

I have debugged the whole process again twice. I saw this for the modified version:

```CreateSecretStorageDialog``` passes ```doBootstrapUIAuth``` to ```authUploadDeviceSigningKeys```
This is used in ```src\rust-crypto\CrossSigningIdentity.ts``` in ```resetCrossSigning```in the line
```await this.outgoingRequestProcessor.makeOutgoingRequest(req, authUploadDeviceSigningKeys);```

which leads to 
```src\rust-crypto\OutgoingRequestProcessor.ts``` with
```typescript
else if (msg instanceof UploadSigningKeysRequest) {
  await this.makeRequestWithUIA(
		  Method.Post,
		  "/_matrix/client/v3/keys/device_signing/upload",
		  {},
		  msg.body,
		  uiaCallback,
  );
  // SigningKeysUploadRequest does not implement OutgoingRequest and does not need to be marked as sent.
  return;
}
```
So afterwards the ```e2e_cross_signing_keys``` table has the three entries (master, self_signing, user_signing).
The next request in the loop of ```resetCrossSigning``` then also sets ```e2e_cross_signing_signatures```.


On the other hand, if I debug it without changes I run in ```bootstrapCrossSigning``` into
```typescript
if (olmDeviceHasKeys) {
	if (!(await this.secretStorage.hasKey())) {
		 logger.warn(
			  "bootstrapCrossSigning: Olm device has private keys, but secret storage is not yet set up; doing nothing for now.",
		 );
```
and it just logged
```
bootstrapCrossSigning: starting {setupNewCrossSigning: undefined, olmDeviceHasMaster: true, olmDeviceHasUserSigning: true, olmDeviceHasSelfSigning: true, privateKeysInSecretStorage: false}
bootstrapCrossSigning: Olm device has private keys, but secret storage is not yet set up; doing nothing for now.
```
I cannot see that anything happens, neither in the code (performing just read operations? ```typescript 
await this.olmMachine.crossSigningStatus()``` and ```typescript await this.secretStorage.get("m.cross_signing....")```) nor in the database and also no SSO in the UI which usually is done when I change anything cross signing key related.

Reply:
> I know that know the implementation waayy better than me

Honestly I've forgotten how a lot of this works, so this is a useful reminder, thank you.

Anyway: the way this is supposed to work is that, on the first call to `bootstrapCrossSigning`, `olmDeviceHasKeys` should be false. So we either load the cross-signing keys from secret storage, or call `resetCrossSigning` (which will create new keys and publish the public keys with `/_matrix/client/v3/keys/device_signing/upload`).

So we need to find out why the OlmDevice has the private keys, and yet the public keys have not been published. https://github.com/element-hq/element-web/issues/27253 might be relevant, but normally the public keys should be published as soon as they are created.